src/graph/entities/attribute_set.c

```diff
@@ -0,0 +1,295 @@
+	AttributeSet set = rm_malloc(sizeof(_AttributeSet));
+
+	set->attr_count = 0;
+
+	return set;
+}
+
+// retrieves a value from set
+// NOTE: if the key does not exist
+//       we return the special constant value ATTRIBUTE_NOTFOUND
+SIValue *AttributeSet_Get
+(
+	const AttributeSet set,  // set to retieve attribute from
+	Attribute_ID attr_id     // attribute identifier
+) {
+	if(set == NULL) return ATTRIBUTE_NOTFOUND;
+
+	if(attr_id == ATTRIBUTE_ID_NONE) return ATTRIBUTE_NOTFOUND;
+
+	// TODO: benchmark, consider alternatives:
+	// sorted set
+	// array divided in two:
+	// [attr_id_0, attr_id_1, attr_id_2, value_0, value_1, value_2]
+	for(int i = 0; i < set->attr_count; i++) {
+		Attribute *attr = set->attributes + i;
+		if(attr_id == attr->id) {
+			// note, unsafe as attribute-set can get reallocated
+			// TODO: why do we return a pointer to value instead of a copy ?
+			// especially when AttributeSet_GetIdx returns SIValue
+			// note AttributeSet_Update operate on this pointer
+			return &attr->value;
+		}
+	}
+
+	return ATTRIBUTE_NOTFOUND;
+}
+
+// retrieves a value from set by index
+SIValue AttributeSet_GetIdx
+(
+	const AttributeSet set,  // set to retieve attribute from
+	int i,                   // index of the property
+	Attribute_ID *attr_id    // attribute identifier
+) {
+	ASSERT(set != NULL);
+	ASSERT(i < set->attr_count);
+	ASSERT(attr_id != NULL);
+
+	Attribute *attr = set->attributes + i;
+	*attr_id = attr->id;
+
+	return attr->value;
+}
+
+// adds an attribute to the set
+void AttributeSet_Add
+(
+	AttributeSet *set,     // set to update
+	Attribute_ID attr_id,  // attribute identifier
+	SIValue value          // attribute value
+) {
+	ASSERT(set != NULL);
+	ASSERT(attr_id != ATTRIBUTE_ID_NONE);
+
+	AttributeSet _set = *set;
+
+	// validate value type
+	// value must be a valid property type
+	ASSERT(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE);
+
+	// make sure attribute isn't already in set
+	ASSERT(AttributeSet_Get(_set, attr_id) == ATTRIBUTE_NOTFOUND);
+
+	// allocate room for new attribute
+	if(_set == NULL) {
+		_set = rm_malloc(sizeof(_AttributeSet) + sizeof(Attribute));
+		_set->attr_count = 1;
+	} else {
+		_set->attr_count++;
+		size_t n = ATTRIBUTESET_BYTE_SIZE(_set);
+		_set = rm_realloc(_set, n);
+	}
+
+	// set attribute
+	Attribute *attr = _set->attributes + _set->attr_count - 1;
+	attr->id = attr_id;
+	attr->value = SI_CloneValue(value);
+
+	// update pointer
+	*set = _set;
+}
+
+// adds or updates an attribute to the set null value allowed
+void AttributeSet_Set_Allow_Null
+(
+	AttributeSet *set,     // set to update
+	Attribute_ID attr_id,  // attribute identifier
+	SIValue value          // attribute value
+) {
+	ASSERT(set != NULL && *set != NULL);
+	ASSERT(attr_id != ATTRIBUTE_ID_NONE);
+
+	AttributeSet _set = *set;
+
+	// validate value type
+	// value must be a valid property type
+	ASSERT(SI_TYPE(value) & (SI_VALID_PROPERTY_VALUE | T_NULL));
+
+	// make sure attribute isn't already in set
```

Comment:
update comment, we do allow for the attribute to exists, in which case we overwrite it.